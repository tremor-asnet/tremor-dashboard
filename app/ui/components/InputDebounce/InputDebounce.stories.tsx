import type { Meta, StoryObj } from "@storybook/react";

// Components
import InputDebounce from "./InputDebounce";

const meta = {
  title: "Components/Common/InputDebounce",
  component: InputDebounce,
  tags: ["autodocs"],
  argTypes: {
    field: { description: "keyword of input search" },
  },
} as Meta<typeof InputDebounce>;

export default meta;

type Story = StoryObj<typeof meta>;

export const InputDebounceDefault: Story = {
  args: {
    field: "Product",
    param: "page",
    valueParam: "1",
  },
};
