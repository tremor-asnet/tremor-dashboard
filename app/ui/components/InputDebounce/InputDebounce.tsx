"use client";

import { useCallback } from "react";
import { usePathname, useRouter, useSearchParams } from "next/navigation";

// Components
import { InputSearch } from "@/ui/components";

// Helpers
import { debounce } from "@/helpers/debounce";

interface InputSearchProps {
  field: string;
  debounceTime?: number;
  param: string;
  valueParam: string;
}

const InputDebounce = ({
  field,
  debounceTime = 1000,
  param,
  valueParam,
}: InputSearchProps) => {
  const searchParams = useSearchParams();
  const pathname = usePathname();
  const { replace } = useRouter();
  const params = new URLSearchParams(searchParams);
  const paramsValue = params.get(field);

  const debouncedHandler = useCallback(
    (
      value: string,
      fieldParam: string,
      pathnameParam: string,
      urlParams: URLSearchParams,
    ) => {
      const handleUrlParams = () => {
        urlParams.set(param, valueParam);

        if (value) {
          urlParams.set(fieldParam, value);
        } else {
          urlParams.delete(fieldParam);
        }

        replace(`${pathnameParam}?${urlParams.toString()}`);
      };

      return debounce(handleUrlParams, debounceTime);
    },
    [replace],
  );

  const handleSearch = (value: string) => {
    debouncedHandler(value, field, pathname, params)();
  };

  const resetSearch = () => {
    params.delete(field);
    replace(`${pathname}?${params.toString()}`);
  };

  return (
    <InputSearch
      onReset={resetSearch}
      onChange={handleSearch}
      value={paramsValue ?? ""}
    />
  );
};

export default InputDebounce;
