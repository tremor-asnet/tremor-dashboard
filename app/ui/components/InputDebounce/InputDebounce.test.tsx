import { RenderResult, fireEvent, render } from "@testing-library/react";

// Components
import InputDebounce from "./InputDebounce";

// Mock the necessary modules
jest.mock("next/navigation", () => ({
  ...jest.requireActual("next/navigation"),
  useRouter: () => ({
    replace: jest.fn(),
  }),
  useSearchParams: jest.fn(() => {
    const params = new URLSearchParams();
    const set = (key: string, value: string) => params.set(key, value);
    const deleteParam = (key: string) => params.delete(key);
    return { params, set, delete: deleteParam };
  }),
}));

describe("InputDebounce component", () => {
  let renderComponent: RenderResult;

  beforeEach(() => {
    jest.useFakeTimers();

    renderComponent = render(
      <InputDebounce field="query" param="page" valueParam="1" />,
    );
  });

  afterEach(() => {
    jest.clearAllTimers();
  });

  it("should matches snapshot", () => {
    const { container } = renderComponent;
    expect(container).toMatchSnapshot();
  });

  it("renders with initial value", () => {
    const component = renderComponent;
    const inputElement = component.getByPlaceholderText("Search...");
    expect(inputElement).toBeTruthy();
  });

  it("updates search value on user input", () => {
    const component = renderComponent;
    const inputElement = component.getByPlaceholderText(
      "Search...",
    ) as HTMLInputElement;
    fireEvent.change(inputElement, { target: { value: "new value" } });
    expect(inputElement.value).toBe("new value");
  });

  it("clears search value on close button click", () => {
    const component = renderComponent;
    const inputElement = component.getByPlaceholderText(
      "Search...",
    ) as HTMLInputElement;
    fireEvent.change(inputElement, { target: { value: "new value" } });
    const closeButton = component.getByTestId("close-button");
    fireEvent.click(closeButton);
    expect(inputElement.value).toBe("");
  });

  it("should update URL parameters correctly when user inputs a value", () => {
    const component = renderComponent;
    const inputElement = component.getByPlaceholderText(
      "Search...",
    ) as HTMLInputElement;
    fireEvent.change(inputElement, { target: { value: "new value" } });
    jest.runAllTimers();
  });

  it("should reset search and update URL parameters when reset is clicked", () => {
    const component = renderComponent;
    const inputElement = component.getByPlaceholderText(
      "Search...",
    ) as HTMLInputElement;
    fireEvent.change(inputElement, { target: { value: "new value" } });
    const resetButton = component.getByTestId("close-button");
    fireEvent.click(resetButton);

    expect(component.container).toMatchSnapshot();
  });

  it("should clear the search value and delete the parameter", () => {
    const component = renderComponent;
    const inputElement = component.getByPlaceholderText(
      "Search...",
    ) as HTMLInputElement;
    fireEvent.change(inputElement, { target: { value: "new value" } });

    // Clear the input value
    fireEvent.change(inputElement, { target: { value: "" } });

    jest.runAllTimers(); // Run timers to execute debounced function
    expect(component.container).toMatchSnapshot();
  });
});
