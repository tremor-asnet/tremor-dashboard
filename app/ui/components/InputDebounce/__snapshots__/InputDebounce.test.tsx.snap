// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputDebounce component should clear the search value and delete the parameter 1`] = `
<div>
  <div
    class="tremor-Flex-root flex w-full flex-row justify-end items-center p-6 relative"
  >
    <div
      class="tremor-TextInput-root relative flex items-center min-w-[10rem] outline-none rounded-tremor-default transition duration-100 border shadow-tremor-input dark:shadow-dark-tremor-input bg-tremor-background hover:bg-tremor-background-muted dark:hover:bg-dark-tremor-background-muted text-tremor-content dark:text-dark-tremor-content border-tremor-border w-auto dark:bg-transparent dark:border-white"
    >
      <input
        class="tremor-TextInput-input w-full bg-transparent focus:outline-none focus:ring-0 border-none text-tremor-default rounded-tremor-default transition duration-100 py-2 text-tremor-content-emphasis dark:text-dark-tremor-content-emphasis [appearance:textfield] [&::-webkit-outer-spin-button]:appearance-none [&::-webkit-inner-spin-button]:appearance-none pr-3 pl-3 placeholder:text-tremor-content dark:placeholder:text-dark-tremor-content"
        data-testid="base-input"
        id="search_order"
        placeholder="Search..."
        style="padding-right: 1rem;"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`InputDebounce component should matches snapshot 1`] = `
<div>
  <div
    class="tremor-Flex-root flex w-full flex-row justify-end items-center p-6 relative"
  >
    <div
      class="tremor-TextInput-root relative flex items-center min-w-[10rem] outline-none rounded-tremor-default transition duration-100 border shadow-tremor-input dark:shadow-dark-tremor-input bg-tremor-background hover:bg-tremor-background-muted dark:hover:bg-dark-tremor-background-muted text-tremor-content dark:text-dark-tremor-content border-tremor-border w-auto dark:bg-transparent dark:border-white"
    >
      <input
        class="tremor-TextInput-input w-full bg-transparent focus:outline-none focus:ring-0 border-none text-tremor-default rounded-tremor-default transition duration-100 py-2 text-tremor-content-emphasis dark:text-dark-tremor-content-emphasis [appearance:textfield] [&::-webkit-outer-spin-button]:appearance-none [&::-webkit-inner-spin-button]:appearance-none pr-3 pl-3 placeholder:text-tremor-content dark:placeholder:text-dark-tremor-content"
        data-testid="base-input"
        id="search_order"
        placeholder="Search..."
        style="padding-right: 1rem;"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`InputDebounce component should reset search and update URL parameters when reset is clicked 1`] = `
<div>
  <div
    class="tremor-Flex-root flex w-full flex-row justify-end items-center p-6 relative"
  >
    <div
      class="tremor-TextInput-root relative flex items-center min-w-[10rem] outline-none rounded-tremor-default transition duration-100 border shadow-tremor-input dark:shadow-dark-tremor-input bg-tremor-background hover:bg-tremor-background-muted dark:hover:bg-dark-tremor-background-muted text-tremor-content dark:text-dark-tremor-content border-tremor-border w-auto dark:bg-transparent dark:border-white"
    >
      <input
        class="tremor-TextInput-input w-full bg-transparent focus:outline-none focus:ring-0 border-none text-tremor-default rounded-tremor-default transition duration-100 py-2 text-tremor-content-emphasis dark:text-dark-tremor-content-emphasis [appearance:textfield] [&::-webkit-outer-spin-button]:appearance-none [&::-webkit-inner-spin-button]:appearance-none pr-3 pl-3 placeholder:text-tremor-content dark:placeholder:text-dark-tremor-content"
        data-testid="base-input"
        id="search_order"
        placeholder="Search..."
        style="padding-right: 1rem;"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
